File: Nebula.DAP/Debugger/BreakpointManager.cs
using Microsoft.Extensions.Logging;
using Nebula.Debugger.Bridge;
using System.Collections.Generic;

namespace Nebula.Debugger.Debugger
{
    public class BreakpointManager
    {
        private readonly ILogger _logger;
        private readonly object _lock = new();

        public IReadOnlyList<BreakpointInformation> FunctionBreakpoints => _functionBreakpoints;
        public IReadOnlyDictionary<string, ConcurrentHashSet<BreakpointInformation>> Breakpoints
            => _breakpoints;

        public readonly List<BreakpointInformation> _functionBreakpoints = [];
        public readonly Dictionary<string, ConcurrentHashSet<BreakpointInformation>> _breakpoints = [];

        public BreakpointManager(ILogger logger)
        {
            _logger = logger;


            _logger.LogInformation($"Breakpoint manager initialized!");
        }

        public void AddFunctionBreakpoint(BreakpointInformation bp)
        {
            // TODO :: Check validity
            lock (_lock)
            {
                _functionBreakpoints.Add(bp);
            }
        }

        public void AddBreakpoint(BreakpointInformation bp)
        {
            // TODO :: Check validity
            lock (_lock)
            {
                if (_breakpoints.TryGetValue(bp.Namespace, out ConcurrentHashSet<BreakpointInformation>? set))
                {
                    set.Add(bp);
                    return;
                }

                _breakpoints[bp.Namespace] =
                [
                    bp
                ];
            }
        }

        public void ClearFunctionBreakpoints()
        {
            // TODO :: Check validity
            lock (_lock)
            {
                _functionBreakpoints.Clear();
            }
        }

        public void ClearBreakpoints()
        {
            lock (_lock)
            {
                _breakpoints.Clear();
            }
        }

        public void ClearBreakpoints(string @namespace)
        {
            lock (_lock)
            {
                if (_breakpoints.TryGetValue(@namespace, out ConcurrentHashSet<BreakpointInformation>? set))
                {
                    set.Clear();
                }
            }

        }
    }
}

File: Nebula.Executor/src/BreakpointManager.cpp
#include "BreakpointManager.h"

using namespace nebula::debugger;

void BreakpointManager::AddFunctionBreakpoint(const BreakpointInformation& breakpointInfo)
{
	bpm_lock lock(m_mutex);
	m_functionBreakpoints.insert(breakpointInfo);
}

void BreakpointManager::AddBreakpoint(const BreakpointInformation& breakpointInfo)
{
	bpm_lock lock(m_mutex);

	auto it = m_breakpoints.find(breakpointInfo.GetNamespace());
	if (it == m_breakpoints.end())
	{
		m_breakpoints.insert(std::make_pair(breakpointInfo.GetNamespace(),
			std::unordered_set<BreakpointInformation>{ breakpointInfo }));

		return;
	}

	it->second.insert(breakpointInfo);
}

void BreakpointManager::ClearFunctionBreakpoints()
{
	bpm_lock lock(m_mutex);
	m_functionBreakpoints.clear();
}

void BreakpointManager::ClearBreakpoints(const std::string& _namespace)
{
	bpm_lock lock(m_mutex);
	m_breakpoints[_namespace].clear();
}

void BreakpointManager::ClearBreakpoints()
{
	bpm_lock lock(m_mutex);
	m_breakpoints.clear();
}

const std::unordered_set<BreakpointInformation>& BreakpointManager::GetFunctionBreakpoints() const
{
	return m_functionBreakpoints;
}

const std::unordered_map<std::string, std::unordered_set<BreakpointInformation>>& BreakpointManager::GetBreakpoints() const
{
	return m_breakpoints;
}
